_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,140 @@
+---
+layout: singlepage-overview
+title: Binary Compatibility for library authors
+
+discourse: true
+permalink: /tutorials/:title.html
+---
+
+## Introduction
+
+A diverse and comprehensive set of libraries is important to any productive software ecosystem. While it is easy to develop and distribute Scala libraries, good library authorship goes far
+beyond just writing code and publishing them.
+
+In this guide, we will cover the important topic of **Binary Compatibility**:
+
+* How binary incompatibility can cause production failures in your applications
+* How library authors can avoid breaking binary compatibility, and/or convey breakages clearly to library users when they happen
+
+Before we start, first we need to understand how code is compiled and executed on the Java Virtual Machine (JVM).
+
+## The JVM execution model
+
+Code compiled to run on the JVM is compiled to a platform-independent format called **JVM bytecode** and stored in **Class File** format (with `.class` extension) and these class files are stored
+in JAR files. The bytecode is what we refer to as the **Binary** format.
+
+When application or library code is compiled, their bytecode invokes named references of classes/methods from their dependencies instead of including the dependencies' actual bytecode
+(unless inlining is explicitly requested). During runtime, the JVM classloader will search through the provided class files for classes/methods referenced by and invoke them.
+
+Let's illustrate with an example:  
+
+We got an application `App` that depends on `A` which itself depends on library `C`. When starting the application we need to provide the class files
+for all of `App`, `A` and `C` (something like `java -cp App.jar:A.jar:C.jar:. MainClass`). If we did not provide `C.jar`, or if we provided a `C.jar` that does not contain certain classes/methods
+which `A` expected to exist in library `C`, we will get an exception when our code attempt to invoke the missing classes/methods.
+
+This is what we call **Binary Incompatibility Errors** - The bytecode interface used for compilation differs and is incompatible with the bytecode provided during runtime.
+
+## What are Evictions, Source Compatibility and Binary Compatibility?
+
+Since the classloader only loads the first match of a class, having multiple versions of the same library in the classpath is redundant. 
+Therefore when deciding which JARs to use for compilation, SBT only selects one version from each library (by default the highest), 
```

Comment:
Uf, I wouldn't mention this. First, it doesn't give a lot of value. Second, it's factually not correct, since it can depend on the resolution tool and the semantics that you use (e.g. Ivy and Maven have different semantics and they are really complex...).

Reply:
Damn, this comment is actually for the next paragraph. I reviewed the document but your last update offsetted it by one line.

Reply:
Yeap I'm aware that there are different semantics between tools, but can you customize the classloader behaviour? My understanding is that resolvers decides the list of JARs to put on the classpath, but classloaders behaves the same (which is why I specifically mentions SBT)

Reply:
> My understanding is that resolvers decides the list of JARs to put on the classpath, but classloaders behaves the same
  
Yeah, and this is correct. What I meant is that the resolution changes the order of the artifacts, and therefore the semantics. I wouldn't mention sbt because sbt is not the only Scala build tool. If I were you, I would create a new section for "evictions" that explains it in a more broad sense, and then I would explain that "build tools usually take care of resolving evictions automatically if they can -- otherwise, they delegate this task to the user, who in a conflict has to choose which library to use".